File: scripts/generate-fence-adversarial.py
#!/usr/bin/env python3
"""Generate the deterministic cross-runtime fencing adversarial corpus.

The generator deliberately uses a tiny specified SplitMix64 implementation
rather than Python's random module, so the same seed is stable across Python
versions and can be replayed from a CI receipt.
"""
from __future__ import annotations

import argparse
import hashlib
import json
import pathlib
import sys
from dataclasses import dataclass
from typing import Any

MASK64 = (1 << 64) - 1
DEFAULT_SEED = 0x4F5245534C4F434B  # ASCII-ish "ORESLOCK"
DEFAULT_RANDOM_CASES = 128
MAX_TOKEN = MASK64


@dataclass
class SplitMix64:
    state: int

    def next(self) -> int:
        self.state = (self.state + 0x9E3779B97F4A7C15) & MASK64
        z = self.state
        z = ((z ^ (z >> 30)) * 0xBF58476D1CE4E5B9) & MASK64
        z = ((z ^ (z >> 27)) * 0x94D049BB133111EB) & MASK64
        return (z ^ (z >> 31)) & MASK64


def digest(label: str) -> str:
    return hashlib.sha256(label.encode("utf-8")).hexdigest()


def request(
    token: int,
    operation: str,
    payload: str,
    *,
    tenant: str = "tenant/acme",
    resource: str = "example/jobs/rebuild",
    holder: str | None = "worker-a",
    lease: str | None = "lease-a",
) -> dict[str, Any]:
    value: dict[str, Any] = {
        "tenantScope": tenant,
        "resourceKey": resource,
        "fencingToken": str(token),
        "operationId": operation,
        "payloadSha256": payload,
    }
    if holder is not None:
        value["holder"] = holder
    if lease is not None:
        value["leaseId"] = lease
    return value


def decision(
    current: dict[str, Any] | None,
    incoming: dict[str, Any],
) -> dict[str, Any]:
    if current is None:
        token = incoming["fencingToken"]
        return {
            "kind": "advanced",
            "shouldApply": True,
            "incomingToken": token,
            "currentToken": token,
            "previousToken": None,
        }
    if (
        current["tenantScope"] != incoming["tenantScope"]
        or current["resourceKey"] != incoming["resourceKey"]
    ):
        return {"error": "identity_mismatch"}
    old = int(current["fencingToken"])
    new = int(incoming["fencingToken"])
    if new > old:
        kind = "advanced"
        apply = True
        current_token = str(new)
    elif new < old:
        kind = "stale"
        apply = False
        current_token = str(old)
    elif (
        current["operationId"] == incoming["operationId"]
        and current["payloadSha256"] == incoming["payloadSha256"]
    ):
        kind = "replay"
        apply = False
        current_token = str(old)
    else:
        kind = "token_reuse"
        apply = False
        current_token = str(old)
    return {
        "kind": kind,
        "shouldApply": apply,
        "incomingToken": str(new),
        "currentToken": current_token,
        "previousToken": str(old),
    }


def token_cases(rng: SplitMix64) -> list[dict[str, Any]]:
    boundaries = [
        0,
        1,
        9,
        10,
        99,
        100,
        (1 << 32) - 1,
        (1 << 53) - 1,
        1 << 53,
        (1 << 53) + 1,
        (1 << 63) - 1,
        1 << 63,
        (1 << 63) + 1,
        MAX_TOKEN - 1,
        MAX_TOKEN,
    ]
    cases = [
        {
            "name": f"valid-token-{index:03d}",
            "value": str(value),
            "expected": {"ok": True, "canonical": str(value)},
        }
        for index, value in enumerate(boundaries)
    ]
    for index in range(16):
        value = rng.next()
        cases.append(
            {
                "name": f"valid-random-token-{index:03d}",
                "value": str(value),
                "expected": {"ok": True, "canonical": str(value)},
            }
        )
    invalid = [
        "",
        "00",
        "01",
        "+1",
        "-1",
        " 1",
        "1 ",
        "\t1",
        "1\n",
        "1.0",
        "1e3",
        "1E3",
        "٠",
        "０",
        "¹",
        "1\u0000",
        "18446744073709551616",
        "99999999999999999999",
        "000000000000000000001",
        "not-a-token",
    ]
    cases.extend(
        {
            "name": f"invalid-token-{index:03d}",
            "value": value,
            "expected": {"ok": False, "error": "invalid_fencing_token"},
        }
        for index, value in enumerate(invalid)
    )
    return cases


def request_cases() -> list[dict[str, Any]]:
    good = digest("request-good")
    base = request(1, "op-1", good)

    def case(
        name: str,
        changes: dict[str, Any],
        *,
        ok: bool = False,
        error: str | None = None,
    ) -> dict[str, Any]:
        incoming = dict(base)
        incoming.update(changes)
        expected: dict[str, Any] = {"ok": ok}
        if error is not None:
            expected["error"] = error
        return {"name": name, "incoming": incoming, "expected": expected}

    return [
        case(
            "valid-ascii-byte-boundaries",
            {
                "tenantScope": "t" * 256,
                "operationId": "o" * 128,
                "holder": "h" * 256,
                "leaseId": "l" * 256,
            },
            ok=True,
        ),
        case(
            "valid-two-byte-utf8-boundaries",
            {
                "tenantScope": "é" * 128,
                "operationId": "é" * 64,
                "holder": "é" * 128,
                "leaseId": "é" * 128,
            },
            ok=True,
        ),
        case(
            "valid-four-byte-utf8-tenant-boundary",
            {"tenantScope": "🙂" * 64},
            ok=True,
        ),
        case("tenant-empty", {"tenantScope": ""}, error="empty_field"),
        case("tenant-ascii-overlong", {"tenantScope": "t" * 257}, error="too_long"),
        case(
            "tenant-two-byte-utf8-overlong",
            {"tenantScope": "é" * 129},
            error="too_long",
        ),
        case(
            "tenant-four-byte-utf8-overlong",
            {"tenantScope": "🙂" * 65},
            error="too_long",
        ),
        case("operation-empty", {"operationId": ""}, error="empty_field"),
        case(
            "operation-ascii-overlong",
            {"operationId": "o" * 129},
            error="too_long",
        ),
        case(
            "operation-two-byte-utf8-overlong",
            {"operationId": "é" * 65},
            error="too_long",
        ),
        case("holder-empty-is-not-absence", {"holder": ""}, error="empty_field"),
        case("holder-overlong", {"holder": "é" * 129}, error="too_long"),
        case("lease-empty-is-not-absence", {"leaseId": ""}, error="empty_field"),
        case("lease-overlong", {"leaseId": "🙂" * 65}, error="too_long"),
        case(
            "payload-too-short",
            {"payloadSha256": "a" * 63},
            error="invalid_payload_sha256",
        ),
        case(
            "payload-too-long",
            {"payloadSha256": "a" * 65},
            error="invalid_payload_sha256",
        ),
        case(
            "payload-uppercase",
            {"payloadSha256": "A" * 64},
            error="invalid_payload_sha256",
        ),
        case(
            "payload-non-hex",
            {"payloadSha256": "g" * 64},
            error="invalid_payload_sha256",
        ),
    ]


def decision_cases(rng: SplitMix64, random_cases: int) -> list[dict[str, Any]]:
    cases: list[dict[str, Any]] = []
    explicit = [
        (None, 0, "first-zero"),
        (0, 1, "zero-to-one"),
        ((1 << 53) - 1, 1 << 53, "js-safe-boundary"),
        (1 << 53, (1 << 53) + 1, "js-unsafe-neighbor"),
        ((1 << 63) - 1, 1 << 63, "signed-boundary"),
        (MAX_TOKEN - 1, MAX_TOKEN, "uint64-maximum"),
        (MAX_TOKEN, MAX_TOKEN - 1, "stale-below-maximum"),
    ]
    for index, (old, new, label) in enumerate(explicit):
        incoming = request(new, f"op-{label}-new", digest(f"{label}-new"))
        current = (
            None
            if old is None
            else request(old, f"op-{label}-old", digest(f"{label}-old"))
        )
        cases.append(
            {
                "name": f"boundary-{index:03d}-{label}",
                "current": current,
                "incoming": incoming,
                "expected": decision(current, incoming),
            }
        )

    for index in range(random_cases):
        current_token = rng.next()
        selector = rng.next() % 7
        if selector == 0:
            incoming_token = current_token
        elif selector == 1:
            incoming_token = (current_token + 1) & MASK64
        elif selector == 2:
            incoming_token = (current_token - 1) & MASK64
        elif selector == 3:
            incoming_token = rng.next()
        elif selector == 4:
            incoming_token = 1 << 53
        elif selector == 5:
            incoming_token = 1 << 63
        else:
            incoming_token = MAX_TOKEN

        op_current = f"op-{index:04d}-{rng.next():016x}"
        payload_current = digest(f"current:{index}:{rng.next()}")
        same_op = bool(rng.next() & 1)
        same_payload = bool(rng.next() & 1)
        op_incoming = (
            op_current
            if same_op
            else f"op-next-{index:04d}-{rng.next():016x}"
        )
        payload_incoming = (
            payload_current
            if same_payload
            else digest(f"incoming:{index}:{rng.next()}")
        )
        current = request(
            current_token,
            op_current,
            payload_current,
            holder=f"worker-{rng.next():016x}",
            lease=f"lease-{rng.next():016x}",
        )
        incoming = request(
            incoming_token,
            op_incoming,
            payload_incoming,
            holder=f"worker-{rng.next():016x}",
            lease=f"lease-{rng.next():016x}",
        )
        if index % 29 == 0:
            current["tenantScope"] = "tenant/other"
        elif index % 31 == 0:
            current["resourceKey"] = "example/jobs/other"
        expected = decision(current, incoming)
        item: dict[str, Any] = {
            "name": f"generated-{index:04d}",
            "current": current,
            "incoming": incoming,
        }
        if "error" in expected:
            item["expectedError"] = expected["error"]
        else:
            item["expected"] = expected
        cases.append(item)
    return cases


def build(seed: int, random_cases: int) -> dict[str, Any]:
    if not 1 <= random_cases <= 100_000:
        raise ValueError("random case count must be in 1..100000")
    rng = SplitMix64(seed & MASK64)
    tokens = token_cases(rng)
    requests = request_cases()
    decisions = decision_cases(rng, random_cases)
    return {
        "$comment": (
            "Deterministic adversarial fencing corpus. Generated; edit "
            "scripts/generate-fence-adversarial.py, not this file."
        ),
        "schema": "ores.locks.fence-adversarial/v1",
        "generator": "splitmix64-v1",
        "seed": f"0x{seed & MASK64:016x}",
        "randomDecisionCases": random_cases,
        "counts": {
            "tokenCases": len(tokens),
            "requestCases": len(requests),
            "decisionCases": len(decisions),
            "total": len(tokens) + len(requests) + len(decisions),
        },
        "tokenCases": tokens,
        "requestCases": requests,
        "decisionCases": decisions,
    }


def smoke_profile(value: dict[str, Any]) -> dict[str, Any]:
    """Keep a small checked-in corpus for direct per-runtime smoke tests."""
    selected_tokens = [
        value["tokenCases"][0],
        value["tokenCases"][8],
        value["tokenCases"][14],
        value["tokenCases"][31],
        value["tokenCases"][32],
        value["tokenCases"][47],
    ]
    selected_requests = [
        value["requestCases"][1],
        value["requestCases"][5],
        value["requestCases"][10],
        value["requestCases"][11],
        value["requestCases"][16],
    ]
    selected_decisions = value["decisionCases"][:8]
    result = dict(value)
    result["profile"] = "smoke"
    result["tokenCases"] = selected_tokens
    result["requestCases"] = selected_requests
    result["decisionCases"] = selected_decisions
    result["counts"] = {
        "tokenCases": len(selected_tokens),
        "requestCases": len(selected_requests),
        "decisionCases": len(selected_decisions),
        "total": len(selected_tokens) + len(selected_requests) + len(selected_decisions),
    }
    return result


def canonical_bytes(value: dict[str, Any]) -> bytes:
    return (
        json.dumps(value, ensure_ascii=False, indent=2, sort_keys=False) + "\n"
    ).encode("utf-8")


def parse_seed(value: str) -> int:
    try:
        parsed = int(value, 0)
    except ValueError as error:
        raise argparse.ArgumentTypeError(f"invalid seed: {value}") from error
    if not 0 <= parsed <= MASK64:
        raise argparse.ArgumentTypeError("seed must fit uint64")
    return parsed


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--seed", type=parse_seed, default=DEFAULT_SEED)
    parser.add_argument("--cases", type=int, default=DEFAULT_RANDOM_CASES)
    parser.add_argument("--output", type=pathlib.Path)
    parser.add_argument("--check", type=pathlib.Path)
    parser.add_argument("--smoke", action="store_true")
    args = parser.parse_args()
    if (args.output is None) == (args.check is None):
        parser.error("provide exactly one of --output or --check")
    corpus = build(args.seed, args.cases)
    if args.smoke:
        corpus = smoke_profile(corpus)
    payload = canonical_bytes(corpus)
    if args.output is not None:
        args.output.parent.mkdir(parents=True, exist_ok=True)
        args.output.write_bytes(payload)
        return 0
    assert args.check is not None
    try:
        actual = args.check.read_bytes()
    except FileNotFoundError:
        print(f"missing generated corpus: {args.check}", file=sys.stderr)
        return 1
    if actual != payload:
        smoke = "--smoke " if args.smoke else ""
        print(
            f"generated corpus drift: rerun {pathlib.Path(__file__).name} "
            f"--seed 0x{args.seed:016x} --cases {args.cases} "
            f"{smoke}--output {args.check}",
            file=sys.stderr,
        )
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/generate-fence-adversarial.mjs
#!/usr/bin/env node

import { createHash } from "node:crypto";
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { dirname, resolve } from "node:path";
import process from "node:process";

const MASK64 = (1n << 64n) - 1n;
const DEFAULT_SEED = 0x2050f3ce5eedn;
const DEFAULT_JSON = "target/adversarial/fence-decision.json";
const PROFILE_ITERATIONS = Object.freeze({ pr: 24, scheduled: 512 });
const PROFILE_STORE_SEQUENCE = Object.freeze({ pr: 128, scheduled: 2048 });
const DIGEST_A = "a".repeat(64);
const DIGEST_B = "b".repeat(64);
const DIGEST_C = "0123456789abcdef".repeat(4);
const VALID_FIELDS = new Set([
  "tenantScope",
  "resourceKey",
  "fencingToken",
  "operationId",
  "payloadSha256",
  "holder",
  "leaseId",
]);

function usage(message) {
  if (message) console.error(message);
  console.error(`usage: generate-fence-adversarial.mjs [options]\n\n` +
    `  --profile pr|scheduled   deterministic corpus size (default: pr)\n` +
    `  --seed <integer>         decimal or 0x-prefixed 64-bit seed\n` +
    `  --output <path>          JSON corpus path\n` +
    `  --receipt <path>         write a machine-readable generation receipt\n` +
    `  --check                  fail unless the checked-in output matches\n`);
  process.exit(message ? 64 : 0);
}

function parseArgs(argv) {
  const options = {
    profile: "pr",
    seed: DEFAULT_SEED,
    output: DEFAULT_JSON,
    receipt: undefined,
    check: false,
  };
  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index];
    switch (arg) {
      case "--profile":
        options.profile = argv[++index];
        break;
      case "--seed":
        options.seed = BigInt(argv[++index]);
        break;
      case "--output":
        options.output = argv[++index];
        break;
      case "--receipt":
        options.receipt = argv[++index];
        break;
      case "--check":
        options.check = true;
        break;
      case "-h":
      case "--help":
        usage();
        break;
      default:
        usage(`unknown argument: ${arg}`);
    }
  }
  if (!Object.hasOwn(PROFILE_ITERATIONS, options.profile)) {
    usage(`unknown profile: ${options.profile}`);
  }
  if (options.seed < 0n || options.seed > MASK64) {
    usage("seed must fit unsigned 64 bits");
  }
  return options;
}

function xorshift64star(seed) {
  let state = seed === 0n ? 0x9e3779b97f4a7c15n : seed;
  return () => {
    state ^= state >> 12n;
    state ^= (state << 25n) & MASK64;
    state ^= state >> 27n;
    state &= MASK64;
    return (state * 0x2545f4914f6cdd1dn) & MASK64;
  };
}

function digest(label) {
  return createHash("sha256").update(label, "utf8").digest("hex");
}

function request(token, operationId, payloadSha256, overrides = {}) {
  const value = {
    tenantScope: "tenant/acme",
    resourceKey: "example/jobs/rebuild",
    fencingToken: token.toString(),
    operationId,
    payloadSha256,
    ...overrides,
  };
  for (const key of Object.keys(value)) {
    if (!VALID_FIELDS.has(key)) throw new Error(`internal unknown request field: ${key}`);
  }
  return value;
}

function expected(kind, incoming, current, previous) {
  return {
    kind,
    shouldApply: kind === "advanced",
    incomingToken: incoming.toString(),
    currentToken: current.toString(),
    previousToken: previous === null ? null : previous.toString(),
  };
}

function decisionCase(name, current, incoming, result) {
  return { name, current, incoming, expected: result };
}

function generatedCaseSet(seed, iterations) {
  const next = xorshift64star(seed);
  const cases = [];
  const boundaries = [
    0n,
    1n,
    9n,
    10n,
    99n,
    100n,
    (1n << 53n) - 1n,
    1n << 53n,
    (1n << 53n) + 1n,
    (1n << 63n) - 1n,
    1n << 63n,
    (1n << 63n) + 1n,
    MASK64 - 1n,
    MASK64,
  ];

  for (const [index, token] of boundaries.entries()) {
    const op = `boundary-first-${index}`;
    const incoming = request(token, op, digest(op));
    cases.push(decisionCase(
      `boundary ${token} advances an absent watermark`,
      null,
      incoming,
      expected("advanced", token, token, null),
    ));

    const replayCurrent = request(token, `boundary-replay-${index}`, DIGEST_A);
    const replayIncoming = request(token, `boundary-replay-${index}`, DIGEST_A, {
      holder: `retry-holder-${index}`,
    });
    cases.push(decisionCase(
      `boundary ${token} exact replay is a no-op`,
      replayCurrent,
      replayIncoming,
      expected("replay", token, token, token),
    ));

    const reuseIncoming = request(token, `boundary-reuse-${index}`, DIGEST_B);
    cases.push(decisionCase(
      `boundary ${token} different operation is token reuse`,
      replayCurrent,
      reuseIncoming,
      expected("token_reuse", token, token, token),
    ));
  }

  for (let index = 1; index < boundaries.length; index += 1) {
    const previous = boundaries[index - 1];
    const incomingToken = boundaries[index];
    cases.push(decisionCase(
      `boundary transition ${previous} to ${incomingToken} advances`,
      request(previous, `boundary-current-${index}`, DIGEST_A),
      request(incomingToken, `boundary-next-${index}`, DIGEST_B),
      expected("advanced", incomingToken, incomingToken, previous),
    ));
  }

  for (let index = 0; index < iterations; index += 1) {
    const left = next();
    let right = next();
    if (right === left) right = (right + 1n) & MASK64;
    const low = left < right ? left : right;
    const high = left < right ? right : left;
    const prefix = `seed-${seed.toString(16)}-${index.toString().padStart(4, "0")}`;
    const currentOp = `${prefix}-current`;
    const nextOp = `${prefix}-next`;
    const replayDigest = digest(`${prefix}:replay`);

    cases.push(decisionCase(
      `${prefix} newer token advances`,
      request(low, currentOp, DIGEST_A),
      request(high, nextOp, DIGEST_B),
      expected("advanced", high, high, low),
    ));
    cases.push(decisionCase(
      `${prefix} older token is stale`,
      request(high, nextOp, DIGEST_B),
      request(low, currentOp, DIGEST_A),
      expected("stale", low, high, high),
    ));
    cases.push(decisionCase(
      `${prefix} equal token exact replay`,
      request(high, currentOp, replayDigest),
      request(high, currentOp, replayDigest, { holder: `${prefix}-retry` }),
      expected("replay", high, high, high),
    ));
    cases.push(decisionCase(
      `${prefix} equal token changed payload`,
      request(high, currentOp, replayDigest),
      request(high, currentOp, digest(`${prefix}:changed`)),
      expected("token_reuse", high, high, high),
    ));
  }

  const utf8Tenant = "é".repeat(128);
  const utf8Operation = "é".repeat(64);
  const utf8Resource = "é".repeat(256);
  const utf8Metadata = "é".repeat(128);
  cases.push(decisionCase(
    "exact UTF-8 byte ceilings remain valid",
    null,
    request(MASK64, utf8Operation, DIGEST_C, {
      tenantScope: utf8Tenant,
      resourceKey: utf8Resource,
      holder: utf8Metadata,
      leaseId: utf8Metadata,
    }),
    expected("advanced", MASK64, MASK64, null),
  ));

  cases.push({
    name: "watermark tenant identity mismatch fails validation",
    current: request(42n, "identity-current", DIGEST_A, { tenantScope: "tenant/other" }),
    incoming: request(43n, "identity-incoming", DIGEST_B),
    expectedError: "identity_mismatch",
  });
  cases.push({
    name: "watermark resource identity mismatch fails validation",
    current: request(42n, "identity-current-resource", DIGEST_A, {
      resourceKey: "example/jobs/other",
    }),
    incoming: request(43n, "identity-incoming-resource", DIGEST_B),
    expectedError: "identity_mismatch",
  });

  return { cases, boundaries };
}


function buildStoreSequence(seed, requestedCount) {
  const next = xorshift64star(seed ^ 0xd1b54a32d192ed03n);
  const sequence = [];
  let current = null;
  let stateValue = null;

  function append(kind, token, operationId, payloadSha256) {
    const previousToken = current?.token ?? null;
    const requestValue = request(token, operationId, payloadSha256);
    const serializedValue =
      `state-${sequence.length}-${digest(`${seed}:${sequence.length}`).slice(0, 20)}`;
    const shouldApply = kind === "advanced";
    if (shouldApply) {
      current = { token, operationId, payloadSha256 };
      stateValue = serializedValue;
    }
    if (current === null || stateValue === null) {
      throw new Error("store sequence must begin with an advanced decision");
    }
    sequence.push({
      request: requestValue,
      serializedValue,
      expected: {
        decision: kind,
        shouldApply,
        currentToken: current.token.toString(),
        previousToken: previousToken === null ? null : previousToken.toString(),
        stateValue,
      },
    });
  }

  const critical = [
    0n,
    1n,
    (1n << 53n) - 1n,
    1n << 53n,
    (1n << 53n) + 1n,
    (1n << 63n) - 1n,
    1n << 63n,
    (1n << 63n) + 1n,
  ];
  for (const [index, token] of critical.entries()) {
    append("advanced", token, `sequence-critical-${index}`, digest(`sequence-critical-${index}`));
    append("replay", token, `sequence-critical-${index}`, digest(`sequence-critical-${index}`));
    append("token_reuse", token, `sequence-critical-reuse-${index}`, digest(`sequence-critical-${index}`));
    if (token > 0n) {
      append("stale", token - 1n, `sequence-critical-stale-${index}`, DIGEST_A);
    }
  }

  const terminalReserve = 8;
  while (sequence.length < requestedCount - terminalReserve) {
    if (current === null) throw new Error("internal sequence state missing");
    const selector = Number(next() % 4n);
    const index = sequence.length;
    if (selector === 0 && current.token < MASK64) {
      const remaining = MASK64 - current.token;
      const delta = 1n + (next() % (remaining < 1_000_000n ? remaining : 1_000_000n));
      const token = current.token + delta;
      append("advanced", token, `sequence-advance-${index}`, digest(`advance:${seed}:${index}`));
    } else if (selector === 1 && current.token > 0n) {
      const token = next() % current.token;
      append("stale", token, `sequence-stale-${index}`, digest(`stale:${seed}:${index}`));
    } else if (selector === 2) {
      append("replay", current.token, current.operationId, current.payloadSha256);
    } else {
      append(
        "token_reuse",
        current.token,
        `sequence-reuse-${index}`,
        digest(`reuse:${seed}:${index}`),
      );
    }
  }

  for (const [terminalIndex, token] of [MASK64 - 1n, MASK64].entries()) {
    const index = critical.length + terminalIndex;
    append("advanced", token, `sequence-critical-${index}`, digest(`sequence-critical-${index}`));
    append("replay", token, `sequence-critical-${index}`, digest(`sequence-critical-${index}`));
    append("token_reuse", token, `sequence-critical-reuse-${index}`, digest(`sequence-critical-${index}`));
    append("stale", token - 1n, `sequence-critical-stale-${index}`, DIGEST_A);
  }

  if (sequence.length !== requestedCount) {
    throw new Error(`store sequence length ${sequence.length} does not match ${requestedCount}`);
  }
  return sequence;
}

function buildCorpus(profile, seed) {
  const { cases, boundaries } = generatedCaseSet(seed, PROFILE_ITERATIONS[profile]);
  const storeSequence = buildStoreSequence(seed, PROFILE_STORE_SEQUENCE[profile]);
  const invalidTokens = [
    "",
    "00",
    "01",
    "+1",
    "-1",
    "1.0",
    " 1",
    "1 ",
    "1e3",
    "1E3",
    "1_000",
    "١",
    "１２",
    "1\u0000",
    "1\n",
    "18446744073709551616",
    "99999999999999999999",
    "not-a-token",
  ];
  const invalidFields = [
    {
      name: "tenant UTF-8 bytes exceed 256",
      field: "tenantScope",
      value: "é".repeat(129),
      expectedError: "too_long",
    },
    {
      name: "resource UTF-8 bytes exceed 512",
      field: "resourceKey",
      value: "é".repeat(257),
      expectedError: "too_long",
    },
    {
      name: "operation UTF-8 bytes exceed 128",
      field: "operationId",
      value: "é".repeat(65),
      expectedError: "too_long",
    },
    {
      name: "holder UTF-8 bytes exceed 256",
      field: "holder",
      value: "é".repeat(129),
      expectedError: "too_long",
    },
    {
      name: "lease UTF-8 bytes exceed 256",
      field: "leaseId",
      value: "é".repeat(129),
      expectedError: "too_long",
    },
    { name: "empty tenant", field: "tenantScope", value: "", expectedError: "empty_field" },
    { name: "empty resource", field: "resourceKey", value: "", expectedError: "empty_field" },
    { name: "empty operation", field: "operationId", value: "", expectedError: "empty_field" },
    { name: "empty holder", field: "holder", value: "", expectedError: "empty_field" },
    { name: "empty lease", field: "leaseId", value: "", expectedError: "empty_field" },
  ];
  const wrongTypeCases = [
    { field: "tenantScope", value: 7, expectedError: "invalid_type" },
    { field: "resourceKey", value: ["example/jobs/rebuild"], expectedError: "invalid_type" },
    { field: "fencingToken", value: 42, expectedError: "invalid_fencing_token" },
    { field: "operationId", value: false, expectedError: "invalid_type" },
    { field: "payloadSha256", value: [DIGEST_A], expectedError: "invalid_payload_sha256" },
    { field: "holder", value: null, expectedError: "invalid_type" },
    { field: "leaseId", value: { value: "lease-1" }, expectedError: "invalid_type" },
  ];

  return {
    $comment: "Deterministic cross-runtime fencing corpus. Generated by scripts/generate-fence-adversarial.mjs; do not edit by hand.",
    schema: "ores.locks-and-leases.fence-corpus/v2",
    profile,
    seed: `0x${seed.toString(16)}`,
    generator: "scripts/generate-fence-adversarial.mjs",
    generatedCaseCount: cases.length,
    tokenBoundaries: boundaries.map(String),
    cases,
    invalidTokens,
    invalidFields,
    wrongTypeCases,
    storeSequence,
  };
}

function sha256(text) {
  return createHash("sha256").update(text, "utf8").digest("hex");
}

async function compareOrWrite(path, rendered, check) {
  const absolute = resolve(path);
  if (!check) {
    await mkdir(dirname(absolute), { recursive: true });
    await writeFile(absolute, rendered, "utf8");
    return { path, state: "written", sha256: sha256(rendered) };
  }
  let actual;
  try {
    actual = await readFile(absolute, "utf8");
  } catch (error) {
    return { path, state: "missing", sha256: sha256(rendered), error: String(error) };
  }
  return {
    path,
    state: actual === rendered ? "matched" : "stale",
    sha256: sha256(rendered),
    actualSha256: sha256(actual),
  };
}

async function main() {
  const options = parseArgs(process.argv.slice(2));
  const corpus = buildCorpus(options.profile, options.seed);
  const json = `${JSON.stringify(corpus, null, 2)}\n`;
  const outputs = [await compareOrWrite(options.output, json, options.check)];
  const passed = outputs.every((output) =>
    output.state === "matched" || output.state === "written");
  const receipt = {
    schema: "ores.locks-and-leases.adversarial-receipt/v1",
    status: passed ? "passed" : "stopped_for_evaluation",
    profile: options.profile,
    seed: corpus.seed,
    generator: corpus.generator,
    generatedCaseCount: corpus.generatedCaseCount,
    invalidTokenCount: corpus.invalidTokens.length,
    invalidFieldCount: corpus.invalidFields.length,
    wrongTypeCount: corpus.wrongTypeCases.length,
    storeSequenceCount: corpus.storeSequence.length,
    outputs,
    exactHead: process.env.GITHUB_SHA ?? null,
    runtime: process.version,
    zeroUnexplainedFindings: passed,
  };
  if (options.receipt) {
    await mkdir(dirname(resolve(options.receipt)), { recursive: true });
    await writeFile(resolve(options.receipt), `${JSON.stringify(receipt, null, 2)}\n`, "utf8");
  }
  console.log(JSON.stringify(receipt));
  if (!passed) process.exitCode = 1;
}

await main();
